### ISSUE TITLE
regression/change: certs now take > 1min to show up (causing lds: fetch failure: network error and ingress tests failure)

### ISSUE THREAD

**Issue by ldemailly** [istio/istio#3757]
```
I0224 14:44:58.148] --- FAIL: TestIngressController (130.95s)
I0224 14:44:58.148] 	kubernetes.go:226: Failed to satisfy function
I0224 14:44:58.149] 	controller_test.go:217: expected 5 IngressRule events to be notified, found 0
I0224 14:44:58.149] 	kubernetes.go:226: Failed to satisfy function
I0224 14:44:58.149] 	controller_test.go:229: expected 5 IngressRule objects to be created, found 0
I0224 14:44:58.149] E0224 14:44:28.134782   29431 event.go:260] Could not construct reference to: '&v1.ConfigMap{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"istio-ingress-controller-leader-istio", GenerateName:"", Namespace:"test", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string{"control-plane.alpha.kubernetes.io/leader":"{\"holderIdentity\":\"istio-ingress-controller-leader-istio\",\"leaseDurationSeconds\":30,\"acquireTime\":\"2018-02-24T14:44:28Z\",\"renewTime\":\"2018-02-24T14:44:28Z\",\"leaderTransitions\":0}"}, OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Data:map[string]string(nil)}' due to: 'selfLink was empty, can't make reference'. Will not report event: 'Normal' 'LeaderElection' 'istio-ingress-controller-leader-istio became leader'
I0224 14:44:58.149] FAIL
I0224 14:44:58.150] FAIL	istio.io/istio/pilot/pkg/config/kube/ingress	161.249s
```
https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3754/istio-unit-tests/1046/

**Comment by ldemailly** [comment]
that both that test and #2199 are having a spike of recent occurrence in both prow and circle ci makes me think there is some actual issue with ingress or envoy/pilot readiness or startup time ?

**Comment by ldemailly** [comment]
hey @wattli @myidpt can you guys dig into this one:

here is what costin and I found looking into the logs of
https://circleci.com/gh/istio/istio/37779#artifacts/containers/0

basically what seemed to happen is that it took 1min30s for the cert to show up

2018-02-26T21:56:35.956655Z	info	watchFileEvents: "/etc/certs//key.pem": CREATE

during which envoy gets

[2018-02-26 21:56:27.531][59][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error

causing
2018-02-26T21:56:31.261250Z	warn	Attempt 6 : http.Get error Get http://10.142.1.30:30109/fortio/debug: dial tcp 10.142.1.30:30109: getsockopt: connection refused

in the tests

**Comment by wattli** [comment]
Is this the 503 issue that Andra talked about?

**Comment by ldemailly** [comment]
no it's a different new issue

**Comment by wattli** [comment]
The file path is wrong "/etc/certs//key.pem", it should be only one '//' in it. Hopefully this is the cause.

 I am not in office today, otherwise we can discuss this tomorrow.

**Comment by ldemailly** [comment]
it's not the cause as it does work after that point - that's just cosmetic issue

**Comment by wattli** [comment]
This actually relates to what Andra mentioned, which is that it takes more than 1 min for ingress to load key/cert. Will look into it tomorrow.

**Comment by ldemailly** [comment]
it's not 503s though, it's different

**Comment by wattli** [comment]
@ldemailly , why do you link #3793 to this issue, is that PR causing the regression?

**Comment by ldemailly** [comment]
because that other pr had that same failure, said that there to explain why the merge despite the failure

**Comment by andraxylia** [comment]
The 503 issue happens continuously, not only at startup.

**Comment by ldemailly** [comment]
@andraxylia let's not mix up the issues. again, this is not about 503s.

**Comment by wattli** [comment]
Can we constantly reproduce this issue? @ldemailly

**Comment by ldemailly** [comment]
sorry I'm not oncall anymore, I do see it still on PRs but I don't have statistics

**Comment by ldemailly** [comment]
it's related to LDS failures too (pilot not available)

**Comment by myidpt** [comment]
Hi, I'm trying to reproduce the failure. It seems to be flaky. Is the following failure related to this issue?
https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/4024/istio-unit-tests/1591/

**Comment by ldemailly** [comment]
you can make a PR where you change (back) simple test ingress from 7*15s retries to 4*10s and that should fail fairly often

**Comment by myidpt** [comment]
@ldemailly I will send another PR to fix the Envoy bootstrapping without cert issue (when auth is enabled).
For your suggestion on bringing up Istio CA before other services, I don't think that's key solution for the issue. For 1), my next PR is going to force Pilot agent to wait for the certs to be ready before bringing up Envoy; 2) Reordering the service sequence in YAML won't help, because the services are brought up in parallel.

**Comment by ldemailly** [comment]
every little bit help; so still having the service account first then CA then pilot probably helps

**Comment by andraxylia** [comment]
I am hitting this for every single run with my local run on minikube. I see with the latest Istio-CA. It happens even when I have auth_enable false.

**Comment by andraxylia** [comment]
It turns out somebody changed the port pilot is listening on from 15003 to 15007 or the other way around without sending a notice, so I ended up with mismatched images and injector.

**Comment by ldemailly** [comment]
can you open a new issue, what you describe seems different from this issue

**Comment by andraxylia** [comment]
In the pilot e2e tests, I moved the istio-ca deployment before istio-pilot, and added a sleep, this should help:
https://github.com/istio/istio/pull/4039/files#diff-caf19981f0e20e07742a823aa01ef35c

**Comment by FuzzOli87** [comment]
I'm also seeing my proxy logs get flooded with these same errors and I'm running istio with mTLS disabled.

**Comment by myidpt** [comment]
@FuzzOli87 What errors are you seeing and what version are you using?

**Comment by FuzzOli87** [comment]
@myidpt 
This is one container:
`[2018-05-14 17:00:40.301][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:00:45.474][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:01:04.533][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:01:10.652][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:03:08.200][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:06:10.007][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:07:13.714][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:07:38.628][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:07:49.121][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:09:11.526][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:09:28.164][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:09:52.381][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:10:40.777][16][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error`

This is another:
`[2018-05-14 16:59:51.858][15][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:00:06.430][15][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:00:12.755][15][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error
[2018-05-14 17:00:33.733][15][info][upstream] external/envoy/source/server/lds_subscription.cc:70] lds: fetch failure: network error`

I'm using istio 0.7.1. mTLS disabled, ingress is up but not exposed(not using it) At first I didn't deploy the ingress but I deployed it after I saw this issue but after bringing it back up and restarting pilot these logs still show.

**Comment by myidpt** [comment]
This error probably complains Envoy is not able to connect to Pilot.
In V0.7.1, without MTLS, the controlPlaneAuthPolicy is by default NONE (unless you explicitly enable MTLS for it). So I don't think this is related to mTLS key/cert not ready. I deployed V0.7.1 on GKE and didn't see this error on ingress or user services. I think may be your Pilot starts too slowly? 

@andraxylia

**Comment by FuzzOli87** [comment]
@myidpt Would it still happen even after Pilot reboots?

**Comment by filipre** [comment]
Is there something we could provide for you making it easier to fix? I believe this error is also responsible for not resolving *some* request to the right service.

**Comment by andraxylia** [comment]
can you please try with 0.8 and let us know if the issue is still there?

**Comment by filipre** [comment]
I can provide feedback once we upgraded our cluster to 0.8. 
However, as it turned out, I was wrong with my believe above. Our internal names/labels were simply too long.

**Comment by filipre** [comment]
upgrading Istio to 0.8 solved the `lds: fetch failure: network error` issue as well.
